File: common/src/main/java/com/study/common/security/InternalServiceRequestWrapper.java
package com.study.common.security;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletRequestWrapper;

import java.util.*;

public class InternalServiceRequestWrapper extends HttpServletRequestWrapper {

    private final Map<String, String> customHeaders = new HashMap<>();

    public InternalServiceRequestWrapper(HttpServletRequest request, String serviceName) {
        super(request);
        customHeaders.put("X-Internal-Service", serviceName);
    }

    @Override
    public String getHeader(String name) {
        String value = customHeaders.get(name);
        if (value != null) return value;
        return super.getHeader(name);
    }

    @Override
    public Enumeration<String> getHeaderNames() {
        Set<String> names = new LinkedHashSet<>(customHeaders.keySet());
        Enumeration<String> original = super.getHeaderNames();
        while (original.hasMoreElements()) {
            names.add(original.nextElement());
        }
        return Collections.enumeration(names);
    }

    @Override
    public Enumeration<String> getHeaders(String name) {
        String value = customHeaders.get(name);
        if (value != null) return Collections.enumeration(List.of(value));
        return super.getHeaders(name);
    }
}

File: common/src/main/kotlin/com/study/common/security/InternalServiceRequestWrapper.kt
package com.study.common.security

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletRequestWrapper
import java.util.*

class InternalServiceRequestWrapper(
    request: HttpServletRequest,
    serviceName: String
) : HttpServletRequestWrapper(request) {

    private val customHeaders = mutableMapOf("X-Internal-Service" to serviceName)

    override fun getHeader(name: String): String? =
        customHeaders[name] ?: super.getHeader(name)

    override fun getHeaderNames(): Enumeration<String> {
        val names = linkedSetOf<String>()
        names.addAll(customHeaders.keys)
        val original = super.getHeaderNames()
        while (original.hasMoreElements()) {
            names.add(original.nextElement())
        }
        return Collections.enumeration(names)
    }

    override fun getHeaders(name: String): Enumeration<String> {
        val value = customHeaders[name]
        if (value != null) return Collections.enumeration(listOf(value))
        return super.getHeaders(name)
    }
}
